pkg/ottl/ottlfuncs/func_parse_csv.go

```diff
@@ -0,0 +1,216 @@
+			return nil, fmt.Errorf("get header: %w", err)
+		}
+
+		headers := strings.Split(headerStr, headerDelimiterString)
+
+		csvReader := csv.NewReader(strings.NewReader(targetStr))
+		csvReader.Comma = delimiter
+		csvReader.FieldsPerRecord = len(headers)
+		csvReader.LazyQuotes = lazyQuotes
+
+		fields, err := csvReadLine(csvReader)
+		if err != nil {
+			return nil, fmt.Errorf("read csv line: %w", err)
+		}
+
+		return csvHeadersMap(headers, fields)
+	}
+}
+
+func parseCSVIgnoreQuotes[K any](target, header ottl.StringGetter[K], delimiter, headerDelimiter rune) ottl.ExprFunc[K] {
+	headerDelimiterString := string([]rune{headerDelimiter})
+	delimiterString := string([]rune{delimiter})
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		targetStr, err := target.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get target: %w", err)
+		}
+
+		headerStr, err := header.Get(ctx, tCtx)
+		if err != nil {
+			return nil, fmt.Errorf("get header: %w", err)
+		}
+
+		headers := strings.Split(headerStr, headerDelimiterString)
+
+		// Ignoring quotes makes CSV parseable with just string.Split
+		fields := strings.Split(targetStr, delimiterString)
+		return csvHeadersMap(headers, fields)
+	}
+}
+
+// csvHeadersMap creates a map of headers[i] -> fields[i].
+func csvHeadersMap(headers []string, fields []string) (pcommon.Map, error) {
+	pMap := pcommon.NewMap()
+	parsedValues := make(map[string]any)
+
+	if len(fields) != len(headers) {
+		return pMap, fmt.Errorf("wrong number of fields: expected %d, found %d", len(headers), len(fields))
+	}
+
+	for i, val := range fields {
+		parsedValues[headers[i]] = val
+	}
+
+	err := pMap.FromRaw(parsedValues)
+	if err != nil {
+		return pMap, fmt.Errorf("create pcommon.Map: %w", err)
+	}
+
+	return pMap, nil
+}
+
+// csvReadLine reads a CSV line from the csv reader, returning the fields parsed from the line.
+// We make the assumption that the payload we are reading is a single row, so we allow newline characters in fields.
+// However, the csv package does not support newlines in a CSV field (it assumes rows are newline separated),
+// so in order to support parsing newlines in a field, we need to stitch together the results of multiple Read calls.
+func csvReadLine(csvReader *csv.Reader) ([]string, error) {
+	lines := make([][]string, 0, 1)
+	for {
+		line, err := csvReader.Read()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+
+		if err != nil && len(line) == 0 {
+			return nil, fmt.Errorf("read csv line: %w", err)
+		}
+
+		lines = append(lines, line)
+	}
+
+	/*
+		This parser is parsing a single value, which came from a single log entry.
+		Therefore, if there are multiple lines here, it should be assumed that each
+		subsequent line contains a continuation of the last field in the previous line.
+
+		Given a file w/ headers "A,B,C,D,E" and contents "aa,b\nb,cc,d\nd,ee",
+		expect reader.Read() to return bodies:
+		- ["aa","b"]
+		- ["b","cc","d"]
+		- ["d","ee"]
+	*/
+
+	joinedLine := lines[0]
```

Comment:
An empty input string or set of headers will trigger an error in the for loop above, right? I'm asking because we need to guarantee the length of this array is at least 1 or we will get a panic.

Reply:
Good catch, I tested this and it did indeed lead to a panic. I've added a guard to error if len(lines) == 0.